### PR TITLE
Tls13 experiment2

### PR DESCRIPTION
**Pull request by ekr**

### CODE REVIEW

File: addons/tls13-comparison-v1/bootstrap.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  return hexCodes.join("");
+}
+async function generateVariate(seed, label) {
+  const hasher = crypto.subtle;
+  const hash = await hasher.digest("SHA-256", new TextEncoder("utf-8").encode(seed + label));
+  let view = new DataView(hash);
+  return view.getUint32(0) / 0xffffffff;
+}
+
+async function startup(addon) {
```
Comment (rhelmer) [inline addons/tls13-comparison-v1/bootstrap.js]: Doesn't need to be `async`

File: addons/tls13-comparison-v1/bootstrap.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+Cu.import("resource://gre/modules/Preferences.jsm");
+Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/TelemetryController.jsm");
+Cu.import("resource://gre/modules/ClientID.jsm");
+
+Cu.importGlobalProperties(["crypto", "TextEncoder"]);
+const DEBUG = false;
+const ENABLE_PROB = .5;
+let gStarted = false;
+const kVERSION_MAX_PREF = "security.tls.version.max";
```
Comment (rhelmer) [inline addons/tls13-comparison-v1/bootstrap.js]: nit: I don't think the `k` prefix is useful, upper-case is a good enough hint.

File: addons/tls13-comparison-v1/bootstrap.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+function debug(msg) {
+    if (DEBUG) {
+        console.log(`TLS 1.3 Test: ${msg}`);
+    }
+}
+
+function disable() {
+  Experiments.instance().disableExperiment("FROM_API");
```
Comment (rhelmer) [inline addons/tls13-comparison-v1/bootstrap.js]: Also not sure if it's necessary to have a `disable` function if it only contains this one line, might as well do the `disableExperiment` one-liner.

File: addons/tls13-comparison-v1/bootstrap.js (hunk 1 above)
Comment (rhelmer) [inline addons/tls13-comparison-v1/bootstrap.js]: Nit: the first optional param to `startup` is usually called `data` (the second optional one being `reason`).

File: addons/tls13-comparison-v1/bootstrap.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  let userprefs = new Preferences();
+  if (userprefs.isSet(kVERSION_MAX_PREF)) {
+    console.log("User has changed TLS max version. Skipping");
+    experiments.setExperimentBranch(id, "skipped");
+    disable();
+    return;
+  }
+  
+  // Seems startup() function is launched twice after install, we're
+  // unsure why so far. We only want it to run once.
```
Comment (rhelmer) [inline addons/tls13-comparison-v1/bootstrap.js]: I'd suggest printing the `reason` being passed to `startup()`, might help to clear up why it's being called twice... I can help look into this too.

**Comment by ekr** [comment]
I tested this by hand, so I'm feeling OK given that it's a one-off. We can ask the testers to do it just to be sure.

I can add a test, but how do you run it?

**Comment by ekr** [comment]
OK, this is revised.

1. I removed the experiments-dependent code.
2. I made the result console.log() instead of debug so that QA could see it.
3. I re-added async to startup because otherwise you can't use await.

@rhelmer, @felipc: PTAL. I think we're done.

**Comment by ekr** [comment]
On Wed, Mar 22, 2017 at 1:36 PM, Felipe Gomes <notifications@github.com>
wrote:

> *@felipc* approved this pull request.
>
> Assuming that you tested that async startup works fine (I never saw it so
> I don't know if it does), this looks OK to me.
>

I did.


> If you end up making more changes you can remove the gStarted safeguard
> because that's specific to experiments. But I wouldn't bother making new
> changes just for that.
>

Thanks

**Comment by ekr** [comment]
@rhelmer I think I am now awaiting your approval

**Comment by rhelmer** [comment]
lgtm - `startup()` won't be called asynchronously by the AddonManager code but having the `async` keyword shouldn't cause any harm.

**Pull request merged**